packages/ws/src/ws/WebSocketManager.ts

```diff
@@ -0,0 +1,269 @@
+	/**
+	 * How long to wait for a shard to connect before giving up
+	 */
+	handshakeTimeout: number | null;
+	/**
+	 * How long to wait for a shard's HELLO packet before giving up
+	 */
+	helloTimeout: number | null;
+	/**
+	 * How long to wait for a shard's READY packet before giving up
+	 */
+	readyTimeout: number | null;
+}
+
+export type WebSocketManagerOptions = RequiredWebsSocketManagerOptions & OptionalWebSocketManagerOptions;
+
+export type ManagerShardEventsMap = {
+	[K in keyof WebSocketShardEventsMap]: [
+		WebSocketShardEventsMap[K] extends [] ? { shardId: number } : WebSocketShardEventsMap[K][0] & { shardId: number },
+	];
+};
+
+export class WebSocketManager extends AsyncEventEmitter<ManagerShardEventsMap> {
+	/**
+	 * The options being used by this manager
+	 */
+	public readonly options: WebSocketManagerOptions;
+
+	/**
+	 * Internal cache for a GET /gateway/bot result
+	 */
+	private gatewayInformation: {
+		data: APIGatewayBotInfo;
+		expiresAt: number;
+	} | null = null;
+
+	/**
+	 * Internal cache for the shard ids
+	 */
+	private shardIds: number[] | null = null;
+
+	/**
+	 * Strategy used to manage shards
+	 * @default SimpleManagerToShardStrategy
+	 */
+	private strategy: IShardingStrategy = new SimpleShardingStrategy(this);
+
+	public constructor(options: RequiredWebsSocketManagerOptions & Partial<OptionalWebSocketManagerOptions>) {
+		super();
+		this.options = { ...DefaultWebSocketManagerOptions, ...options };
+	}
+
+	public setStrategy(strategy: IShardingStrategy) {
+		this.strategy = strategy;
+		return this;
+	}
+
+	/**
+	 * Fetches the gateway information from Discord - or returns it from cache if available
+	 * @param force Whether to ignore the cache and force a fresh fetch
+	 */
+	public async fetchGatewayInformation(force = false) {
+		if (this.gatewayInformation) {
+			if (this.gatewayInformation.expiresAt <= Date.now()) {
+				this.gatewayInformation = null;
+			} else if (!force) {
+				return this.gatewayInformation.data;
+			}
+		}
+
+		const data = (await this.options.rest.get(Routes.gatewayBot())) as RESTGetAPIGatewayBotResult;
+
+		this.gatewayInformation = { data, expiresAt: Date.now() + data.session_start_limit.reset_after };
+		return this.gatewayInformation.data;
+	}
+
+	/**
+	 * Updates your total shard count on-the-fly, spawning shards as needed
+	 * @param shardCount The new shard count to use
+	 */
+	public async updateShardCount(shardCount: number | null) {
+		await this.strategy.destroy();
+		this.options.shardCount = shardCount;
+
+		const shardIds = await this.getShardIds(true);
+		await this.strategy.spawn(shardIds);
+
+		return this;
+	}
+
+	/**
+	 * Yields the total number of shards across for your bot, accounting for Discord recommendations
+	 */
+	public async getShardCount(): Promise<number> {
+		if (this.options.shardCount) {
+			return this.options.shardCount;
+		}
+
+		const shardIds = await this.getShardIds();
+		return shardIds.length;
```

Comment:
This isn't exactly correct. If I have shard ids `[1, 3, 7]`, shardCount wouldn't be 3 but Math.max(...ids) + 1